File: 2024/day25.go
package main

import (
	"strings"

	"lukechampine.com/advent/utils"
)

var input = utils.Input(2024, 25)

func main() {
	groups := strings.Split(input, "\n\n")
	var locks [][5]int
	var keys [][5]int
	for _, group := range groups {
		lines := utils.Lines(group)
		cols := utils.Transpose(lines).([]string)
		var a [5]int
		for i := range a {
			a[i] = strings.Count(cols[i], "#")
		}
		if utils.CharCounts(lines[0])['#'] == 5 {
			locks = append(locks, a)
		} else {
			keys = append(keys, a)
		}
	}

	var n int
	for _, l := range locks {
		for _, k := range keys {
			n += utils.BoolToInt(utils.All(5, func(i int) bool {
				return l[i]+k[i] <= 7
			}))
		}
	}
	utils.Println(n)
}

File: 2024/rust/src/day25.rs
use crate::utils;

pub fn part1(input: &str) -> String {
    let mut locks: Vec<[usize; 5]> = Vec::new();
    let mut keys: Vec<[usize; 5]> = Vec::new();
    input.split("\n\n").for_each(|g| {
        let a = std::array::from_fn(|i| {
            g.lines()
                .filter(|line| line.bytes().nth(i) == Some(b'#'))
                .count()
        });
        if g.starts_with("#####") {
            locks.push(a);
        } else {
            keys.push(a);
        }
    });
    utils::pairs(&locks, &keys)
        .iter()
        .filter(|(l, k)| (0..5).all(|i| l[i] + k[i] <= 7))
        .count()
        .to_string()
}

pub fn part2(_input: &str) -> String {
    "Merry Christmas!".to_string()
}
